File: day05/day05.rb
module MergeableRanges
  refine Range do
    def overlapping?(other) = min <= other.max && other.min <= max

    def merge(other) = ([min, other.min].min..[max, other.max].max)
  end
end

module Day05
  module_function

  using MergeableRanges

  def solve_one(...)
    parse(...) => ranges:, ingredients:

    ingredients.count { |ingredient| ranges.any? { it.cover?(ingredient) } }
  end

  def solve_two(...)
    parse(...) => ranges:
    merged_ranges = ranges.sort_by(&:min).each_with_object([]) do |curr, merged|
      curr = curr.merge(merged.pop) if merged[-1]&.overlapping?(curr)

      merged << curr
    end

    merged_ranges.sum(&:size)
  end

  def parse(input_filename)
    raw_ranges, raw_ingredients = File.read(input_filename).split("\n\n").map(&:split)
    ranges = raw_ranges.map do |range|
      start, fin = range.split("-").map(&:to_i)
      start..fin
    end

    { ranges:, ingredients: raw_ingredients.map(&:to_i) }
  end
end

File: day05/day05.exs
defmodule Day05 do
  def solve_one(filename) do
    %{ranges: ranges, ingredients: ingredients} = parse(filename)

    Enum.count(ingredients, fn ingredient ->
      Enum.any?(ranges, &(ingredient in &1))
    end)
  end

  def solve_two(filename) do
    %{ranges: ranges} = parse(filename)

    Enum.sort_by(ranges, &Enum.min/1)
    |> Enum.reduce([], &merge_ranges/2)
    |> Enum.sum_by(&Range.size/1)
  end

  def parse(filename) do
    [raw_ranges, raw_ingredients] = File.read!(filename)
      |> String.split("\n\n")
      |> Enum.map(&String.split/1)

    ranges = Enum.map(raw_ranges, fn range ->
      [start, stop] = String.split(range, "-") |> Enum.map(&String.to_integer/1)
      start..stop
    end)

    %{ranges: ranges, ingredients: Enum.map(raw_ingredients, &String.to_integer/1)}
  end

  defp merge_ranges(curr, []), do: [curr]
  defp merge_ranges(curr, [previous | rest]) do
    if not Range.disjoint?(curr, previous) do
      [merge_range(curr, previous) | rest]
    else
      [curr, previous | rest]
    end
  end

  defp merge_range(a, b), do: min(a.first, b.first)..max(a.last, b.last)
end
